### ISSUE TITLE
Firebase Auth linkWithGoogleCredential is broken since about 0.6.2+1

### ISSUE THREAD

**Issue by BerndWessels** [flutter/flutter#25250]
## Steps to Reproduce

 `await linkWithGoogleCredential`  gets stuck - this used to work up to `0.5.20`
 
In the example below it never reaches log step `---------------------- 4`, BUT is also doesn't catch any exception.

```dart
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final GoogleSignIn _googleSignIn = GoogleSignIn();

    GoogleSignInAuthentication googleAuth;
    try {
      print("---------------------- 1");
      final GoogleSignInAccount googleUser = await _googleSignIn.signIn();
      print("---------------------- 2");
      googleAuth = await googleUser.authentication;
      print("---------------------- 3");
      final FirebaseUser user = await _auth.linkWithGoogleCredential(
        accessToken: googleAuth.accessToken,
        idToken: googleAuth.idToken,
      );
      print("---------------------- 4");
    } catch (err) {
      print("---------------------- 5");
      print(err);
    }
```

## Logs

```yaml
  # Firebase
  firebase_auth: ^0.6.6
  firebase_core: ^0.2.5+1
  firebase_dynamic_links: ^0.1.0+1
  cloud_firestore: ^0.8.2+3

  # Social login providers
  flutter_facebook_login: ^1.1.1
  google_sign_in: ^3.2.4
```

```
[√] Flutter (Channel stable, v1.0.0, on Microsoft Windows [Version 10.0.17134.407], locale en-NZ)
    • Flutter version 1.0.0 at C:\Users\b\git\flutter
    • Framework revision 5391447fae (12 days ago), 2018-11-29 19:41:26 -0800
    • Engine revision 7375a0f414
    • Dart version 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)

[√] Android toolchain - develop for Android devices (Android SDK 28.0.3)
    • Android SDK at C:\Users\b\AppData\Local\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = C:\Users\b\AppData\Local\Android\sdk
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[√] Android Studio (version 3.2)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 31.1.1
    • Dart plugin version 181.5656
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[√] IntelliJ IDEA Community Edition (version 2017.3)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2017.3
    • Flutter plugin version 19.1
    • Dart plugin version 173.3727.108

[√] Connected device (1 available)
    • Nexus 6P • ENU7N15A27004986 • android-arm64 • Android 8.1.0 (API 27)

• No issues found!
```

```gradle
buildscript {
    ext.kotlin_version = '1.3.11'
    repositories {
        google()
        jcenter()
        maven {
            url 'https://maven.fabric.io/public'
        }
        maven {
            url 'https://dl.bintray.com/android/android-tools'
        }
    }

    dependencies {
        classpath 'com.android.tools.build:gradle:3.2.1'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        classpath 'com.google.gms:google-services:4.2.0' // Google Services plugin
    }
}
```

**Comment by BerndWessels** [comment]
OK, here seems to be the problem:

`linkWithGoogleCredential` does not work if another firebase account is already linked to the google account.

Unfortunately there is no exception thrown, it is just stuck in the `await` which is a big problem for the program flow.

How do we get to into this situation - easy:

1. user starts app and gets an anonymous firebase user created
2. user signs in with his google account
2.1. anonymous firebase user gets linked with the google account

3. user signs out later
3.1. the app creates another anonymous user for the signed out user
3.2. the old google user is still in firebase and linked to the google account

4. now the user decides to sign in with his google account again
4.1. `linkWithGoogleCredential` fails/get stuck because the google account is already linked to the old firebase user account


This is a pretty common scenario in a lot of apps - is there a workaround or can fix `linkWithGoogleCredential` so that it will not get stuck and either succeed or at least throw an exception?

**Comment by BerndWessels** [comment]
One more common scenario that is also broken by this bug is cross device app access.

Basically a user has the same app on multiple of his devices. He wants to sign on with his google account on all of these devices to the same app.

It will succeed on the first device but on all other devices `linkWithGoogleCredential` will fail as described.

@zoechi I believe this is a really serious issue and needs actually some `firebase auth` redesign or at least rethink - because I can't see a workaround without allowing the same social user linked to multiple firebase users.

**Comment by thevzurd** [comment]
Eagerly waiting for a workaround. I am facing a similar issue here.
![image](https://user-images.githubusercontent.com/31169070/49982670-cc3c9880-ff2b-11e8-9122-343ecd9d70cd.png)

I also tried to used the 'AuthCredential' and GoogleAuthProvider classes mention in the example code, but those classes does not exist in the 'package:firebase_auth/firebase_auth.dart', even though I am using the latest package.

**Comment by MarcinusX** [comment]
Had the same issue. It seems to be fixed in version `8.0.0` as I cannot reproduce it anymore :)

**Comment by zoechi** [comment]
@BerndWessels can you still reproduce with `0.8.0`?

**Comment by BerndWessels** [comment]
@zoechi @MarcinusX still badly broken :(

I've upgraded to

```yaml
  # Firebase
  firebase_auth: 0.8.0+1
  firebase_core: 0.3.0
  firebase_dynamic_links: ^0.1.0+1
  cloud_firestore: 0.9.0

  # Social login providers
  flutter_facebook_login: ^1.1.1
  google_sign_in: ^3.2.4
```

to do so I had to go through all [this pain](https://github.com/flutter/flutter/issues/27128#issuecomment-457918930)

after upgrading the code has changed to `linkWithCredential`:

```dart
  final FirebaseAuth _auth = FirebaseAuth.instance;
  final GoogleSignIn _googleSignIn = GoogleSignIn();

    GoogleSignInAuthentication googleAuth;
    try {
      print("---------------------- 1");
      final GoogleSignInAccount googleUser = await _googleSignIn.signIn();
      print("---------------------- 2");
      googleAuth = await googleUser.authentication;
      final credential = GoogleAuthProvider.getCredential(idToken: googleAuth.idToken, accessToken: googleAuth.accessToken);
      print("---------------------- 3");
      final FirebaseUser user = await _auth.linkWithCredential(credential);
      print("---------------------- 4");
    } catch (err) {
      print("---------------------- 5");
      print(err);
    }
```

Now we get this error:

```
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): Failed to handle method call
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): java.lang.IllegalArgumentException: Given String is empty or null
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at com.google.android.gms.common.internal.Preconditions.checkNotEmpty(Unknown Source:5)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at com.google.firebase.auth.EmailAuthProvider.getCredential(Unknown Source:2)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at io.flutter.plugins.firebaseauth.FirebaseAuthPlugin.handleLinkWithEmailAndPassword(FirebaseAuthPlugin.java:272)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at io.flutter.plugins.firebaseauth.FirebaseAuthPlugin.onMethodCall(FirebaseAuthPlugin.java:122)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:200)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at io.flutter.view.FlutterNativeView.handlePlatformMessage(FlutterNativeView.java:163)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at android.os.MessageQueue.next(MessageQueue.java:325)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at android.os.Looper.loop(Looper.java:142)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at android.app.ActivityThread.main(ActivityThread.java:6494)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
E/MethodChannel#plugins.flutter.io/firebase_auth(23208): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
I/flutter (23208): ---------------------- 5
I/flutter (23208): PlatformException(error, Given String is empty or null, null)
```

I'm starting to doubt that firebase is mature enough for production, or am I just to stupid to see what is going on?

**Comment by zoechi** [comment]
Might be a duplicate of #26477

**Comment by diver2** [comment]
I'm getting the same error as @BerndWessels (last post above)
firebase auth updated to 0.8.1
basic scenario. I got a user that authed with google. No other platform.
I make this user try email/password login, with the same email as the google mail.
I get EMAIL_ALREADY_USED error. I make user log with google - so far so good. The laaaast step, getting the accounts to link, I get the exact same error (see below my prnt - just in case there are slight differences I did not notice: 2:30am here, been more than 1h looking after this... :(

```
I/flutter (31243): LINKIN PARK!
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): Failed to handle method call
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): java.lang.IllegalArgumentException: Given String is empty or null
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at com.google.android.gms.common.internal.Preconditions.checkNotEmpty(Unknown Source)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at com.google.firebase.auth.EmailAuthProvider.getCredential(Unknown Source)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at io.flutter.plugins.firebaseauth.FirebaseAuthPlugin.handleLinkWithEmailAndPassword(FirebaseAuthPlugin.java:272)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at io.flutter.plugins.firebaseauth.FirebaseAuthPlugin.onMethodCall(FirebaseAuthPlugin.java:122)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at io.flutter.plugin.common.MethodChannel$IncomingMethodCallHandler.onMessage(MethodChannel.java:200)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at io.flutter.view.FlutterNativeView.handlePlatformMessage(FlutterNativeView.java:163)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at android.os.MessageQueue.nativePollOnce(Native Method)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at android.os.MessageQueue.next(MessageQueue.java:323)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at android.os.Looper.loop(Looper.java:136)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at android.app.ActivityThread.main(ActivityThread.java:6682)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at java.lang.reflect.Method.invoke(Native Method)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1520)
E/MethodChannel#plugins.flutter.io/firebase_auth(31243): 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1410)
I/flutter (31243): PlatformException(error, Given String is empty or null, null) 
```

**Comment by diver2** [comment]
just, this is Android only... iOS worked on simulator...

**Comment by diver2** [comment]
@zoechi If this is a duplicate, I can't understand what is the fix on #26477 - that is for phone, I'm not coming near phones... just email/password, google...
Am I missing something? Would greatly love some tips... tks

**Comment by diver2** [comment]
Voila! I GOT IT! the fix from kika works  #1069.
Tomorrow I post in detail step by step what I did, if anyone is a newbie like me trying to follow through... but it works... have to implement https://github.com/flutter/plugins/pull/1069/commits/8cb8984360f92f1cebdeae6c169c685781c9a418 on the java files on your project...

**Comment by diver2** [comment]
So, for others that arrive at this thread, if you notice when you click on the https://github.com/flutter/plugins/pull/1069 - there are tabs on top. One of these tabs is market "Commits". 
If you click on the numbers on the side of the commits (the most important one is the first, 7b866c7 - the 2nd is style) - you are gonna see the JAVA code...
Ok, we dev in flutter. How to arrive at java code.
Open Android Studio, open your project at ProjectName/android - wait for it to fully open. Do not update gradle, or do other stuff that android studio suggests.
when it finishes loading, there is a tree structure, with folders: 
- app,
- firebase_core
- firebase_auth
etc

go to firebase_auth, follow the path to firebase_auth/java/io_... /FirebaseAuthPlugin - this is the file you got to edit, just like in the commit. go to the lines of the commit.
Java is not so different from dart, so you are going to understand it. Edit file. Save.
Go back to your flutter project. build it. Happy times.

**Comment by kika** [comment]
@diver2  I'm happy it helped, but c'mon Flutter team, the changes are so trivial, why not to incorporate them into the main code? I understand someone could feel hurt by merging a pull request from some stranger on the internet, but you can just rewrite these changes. It's just a few lines of code.

**Comment by kika** [comment]
Also, the way @diver2 suggests to incorporate changes will work, but you will change the plugin files directly where they are pulled in by `flutter pub get`, so you may eventually overwrite your changes with unpatched versions again. The more robust way is to clone my [fork](https://github.com/kika/plugins.git) somewhere:

**Comment by tapkain** [comment]
To anybody wondering how to fix the null email crash. The reason of this crash is that in the firebase_auth.dart, linkWithCredentials we are marshalling "data" hash map to java.
In java, FirebaseAuthPlugin.java, handleLinkWithEmailAndPassword we are getting the arguments as "email" and "password" but they are boxed inside the data hash map property, so we need to unbox this property and after that read the email password fields.

TL;DR
FirebaseAuthPlugin.java, handleLinkWithEmailAndPassword
`private void handleLinkWithEmailAndPassword(
      MethodCall call, Result result, FirebaseAuth firebaseAuth) {
    final FirebaseUser currentUser = firebaseAuth.getCurrentUser();
    if (currentUser == null) {
      markUserRequired(result);
      return;
    }

    Map<String, Map<String, String>> arguments = call.arguments();
    Map<String, String> data = arguments.get("data");

    String email = data.get("email");
    String password = data.get("password");

    AuthCredential credential = EmailAuthProvider.getCredential(email, password);

    currentUser
        .linkWithCredential(credential)
        .addOnCompleteListener(new SignInCompleteListener(result));
  }`

**Comment by collinjackson** [comment]
I think this should be fixed as of firebase_auth 0.8.2.

flutter/plugins#1333

**Comment by kroikie** [comment]
@BerndWessels

This issue has been moved to https://github.com/FirebaseExtended/flutterfire/issues/549. Any further collaboration will be done there.
